src/ucp/wireup/select.c

```diff
@@ -453,6 +460,64 @@ ucp_wireup_select_transport(ucp_ep_h ep, const ucp_address_entry_t *address_list
     return UCS_OK;
 }
 
+/* Return:
+ * - UCS_OK if suitable transport was found
+ * - UCS_INPROGRESS if suitable transport wasn't found, but
+ *   it can be replaced by emulation over AM
+ * - other error code if an error occured during selection */
+static ucs_status_t
+ucp_wireup_select_lane(ucp_wireup_select_ctx_t *select_ctx,
+                       const ucp_wireup_criteria_t *criteria,
+                       uint64_t tl_bitmap, uint64_t remote_md_map,
+                       uint64_t local_dev_bitmap, uint64_t remote_dev_bitmap,
+                       uint32_t usage, int show_error,
+                       ucp_wireup_select_info_t *select_info)
+{
+    ucp_ep_h ep                 = select_ctx->ep;
+    uint64_t scalable_tl_bitmap = (select_ctx->scalable_tl_bitmap &
+                                   tl_bitmap);
+    ucs_status_t status;
+
+    /* Select lane among scalable transports */
+    if (scalable_tl_bitmap) {
+        status = ucp_wireup_select_transport(ep, select_ctx->address_list,
+                                             select_ctx->address_count, criteria,
+                                             scalable_tl_bitmap, remote_md_map,
+                                             local_dev_bitmap, remote_dev_bitmap,
+                                             0 /* don't log errors */, select_info);
+        if (status == UCS_OK) {
+            ucs_assert(ucp_is_scalable_transport(ep->worker->context,
+                                                 select_info->max_num_eps));
+            return status;
+        }
+
+        /* If we get there, it means that the selected AM transport
+         * is scalable enough and could be used instead of more
+         * suitable transport for current lane, but it will meet
+         * scalability requirements */
+        if (/* can't emulate aux over AM */
+            (usage != UCP_WIREUP_LANE_USAGE_AUX) &&
+            /* can we use emulation over AM? */
+            (select_ctx->allow_am ||
+             !(usage & (UCP_WIREUP_LANE_USAGE_RMA    |
```

Comment:
what is checked? If AM is not allowed, but we have RMA lane, then we will propose to use AM emulation?